### PR TITLE
IWeakReferenceSource support

### PR DESCRIPTION
**Pull request by tristanlabelle**
Implements `IWeakReferenceSource` on the COM interface that we synthesize for Swift objects sent to WinRT. This allows WinUI to keep weak references to our Swift objects. Also adds tests.

### CODE REVIEW

File: swiftwinrt/Resources/Support/IWeakReferenceSource.swift [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+fileprivate var IWeakReferenceSourceVTable: C_IWeakReferenceSourceVtbl = .init(
+    QueryInterface: { pUnk, riid, ppvObject in
+        guard let pUnk, let riid, let ppvObject else { return E_INVALIDARG }
+        switch riid.pointee {
+        case IID_IWeakReferenceSource:
+            _ = pUnk.pointee.lpVtbl.pointee.AddRef(pUnk)
+            return S_OK
+        default:
+            guard let obj = WeakReferenceSourceWrapper.tryUnwrapFromBase(raw: pUnk) else { return E_NOINTERFACE }
+            fatalError("\(#function): Not implemented: something like obj.QueryInterface")
```
Comment (tristanlabelle) [inline swiftwinrt/Resources/Support/IWeakReferenceSource.swift:47]: @stevenbrix , I'm blocked here. I think I need to have an `-Impl` object that keeps a COMRef to the object somehow?

File: swiftwinrt/Resources/Support/IWeakReference.swift [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+fileprivate var IWeakReferenceVTable: C_IWeakReferenceVtbl = .init(
+    QueryInterface: { pUnk, riid, ppvObject in
+        guard let pUnk, let riid, let ppvObject else { return E_INVALIDARG }
+        switch riid.pointee {
+        case IID_IWeakReference:
+            _ = pUnk.pointee.lpVtbl.pointee.AddRef(pUnk)
+            return S_OK
+        default:
+            guard let obj = WeakReferenceWrapper.tryUnwrapFromBase(raw: pUnk) else { return E_NOINTERFACE }
+            fatalError("\(#function): Not implemented: something like obj.QueryInterface")
```
Comment (tristanlabelle) [inline swiftwinrt/Resources/Support/IWeakReference.swift:54]: @stevenbrix , I'm blocked the same as for `IWeakReferenceSource`

File: swiftwinrt/Resources/Support/IWeakReference.swift [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+fileprivate func Resolve(
+        _ this: UnsafeMutablePointer<C_IWeakReference>?,
+        _ iid: UnsafePointer<GUID_Workaround>?,
+        _ inspectable: UnsafeMutablePointer<UnsafeMutablePointer<C_IInspectable>?>?) -> HRESULT {
+    guard let weakReference = WeakReferenceWrapper.tryUnwrapFrom(abi: ComPtr(this)) else { return E_FAIL }
+    guard let iid, let inspectable else { return E_INVALIDARG }
+    if let target = weakReference.target {
+        _ = target
+        _ = iid
+        fatalError("\(#function): Not implemented: something like target.QueryInterface")
```
Comment (tristanlabelle) [inline swiftwinrt/Resources/Support/IWeakReference.swift:71]: @stevenbrix , similar to other places I'm blocked, how would I preserve a COM object to this point?

**Review by tristanlabelle — COMMENTED** [review]
@stevenbrix  I think I have most of the implementation but I'm blocked at this point. When I start going down the WeakReferenceSource creation, I don't have a COM object anymore and eventually I need it again for queryInterface implementations. I don't understand how it works in this codebase.

File: swiftwinrt/Resources/Support/IWeakReference.swift (hunk 3 above)
Comment (stevenbrix) [reply swiftwinrt/Resources/Support/IWeakReference.swift:71]: something like this i think:
```
let wrapper = AnyWrapper(target)
wrapper.copyTo(iid, iinspectable)
```

File: swiftwinrt/Resources/Support/IWeakReferenceSource.swift (hunk 1 above)
Comment (tristanlabelle) [reply swiftwinrt/Resources/Support/IWeakReferenceSource.swift]: No but I could piggyback on the `AnyWrapper.queryInterface` implementation.

File: swiftwinrt/Resources/Support/IWeakReference.swift (hunk 2 above)
Comment (tristanlabelle) [reply swiftwinrt/Resources/Support/IWeakReference.swift]: There's is no such method, but I could piggyback on the `AnyWrapper.queryInterface` implementation.

File: swiftwinrt/Resources/Support/IWeakReference.swift (hunk 3 above)
Comment (tristanlabelle) [reply swiftwinrt/Resources/Support/IWeakReference.swift]: Something like that is how I got this to work, thanks!

**Review by stevenbrix — APPROVED** [review]
nice! does this change make it possible to add a public `WeakReference<T: IWinRTObject>` so that we can keep WinRT weak references to objects?

**Comment by tristanlabelle** [comment]
> nice! does this change make it possible to add a public `WeakReference<T: IWinRTObject>` so that we can keep WinRT weak references to objects?

@stevenbrix No, I think that implementation would look quite different, and likely much simpler.

**Pull request merged**